### ISSUE TITLE
Please don't use Boost as testing library and try to lower entry barrier for cpp

### ISSUE THREAD

**Issue by agauniyal** [exercism/exercism.io#2407]
I had installed cli just a few minutes ago and was testing the very first cpp problem that involved testing different cases for bob::hey("sdasd");

First problem that I encountered was managing different boost libraries in my system since cmake complained of not finding them. After some searching , I had to install another package and then ln -s them. Then Cmake was happy , atleast.

I began coding , took some spins on ideone to verify my code before finally writing it in bob.cpp and bob.h. Then I remembered , I can only submit bob.cpp since exercism.io website states some issue submitting 2 or more files.

I cut-pasted stuff from bob.h to bob.cpp and included it in bob_test.cpp so that everything works. Keep in mind I've yet not generated make files , niether executed make.

Now after doing all this stuff , and making sure my code runs on ideone , I tried executing cmake and make. And I was presented with this. And I dare you , all I was missing was a single `}` but I didn't deserved all of this : [error](https://gist.github.com/agauniyal/37b23320dd624b20941b).

and believe me that's not even 30% of what was being outputted on my screen. This is maximum amount of text I've allowed on my terminal.

I would just recommend moving testing excercises that use Boost Library a little forward and writing short manual test cases in some excersises.

**Comment by agauniyal** [comment]
and now after inserting that required brace , I've error as : 

    [ 50%] Building CXX object CMakeFiles/bob.dir/bob_test.cpp.o
    [100%] Building CXX object CMakeFiles/bob.dir/bob.cpp.o
    Linking CXX executable bob
    CMakeFiles/bob.dir/bob.cpp.o: In function `bob::hey(std::string)':
    bob.cpp:(.text+0x79): multiple definition of `bob::hey(std::string)'
    CMakeFiles/bob.dir/bob_test.cpp.o:bob_test.cpp:(.text+0x79): first defined here
    collect2: error: ld returned 1 exit status
    CMakeFiles/bob.dir/build.make:113: recipe for target 'bob' failed
    make[2]: *** [bob] Error 1
    CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/bob.dir/all' failed
    make[1]: *** [CMakeFiles/bob.dir/all] Error 2
    Makefile:75: recipe for target 'all' failed
    make: *** [all] Error 2

and I again don't have any idea how to correct it. I do not have multiple definitions of bob declared at first place , and niether is this error message telling me 2nd place.

**Comment by kytrinyx** [comment]
@LegalizeAdulthood Would you take a look at this one, please?

**Comment by LegalizeAdulthood** [comment]
@agauniyal In order to help you out I need to understand where you are coming from first.  How much experience do you have with C++?  Are you attempting to learn C++ at the same time as starting out with exercism?

**Comment by agauniyal** [comment]
@LegalizeAdulthood I am familiar with cpp since 2+ yrs ( currently in university), but I'm relatively new to cpp14+ stuff. So I'm kind of learning cpp but not from scratch.

**Comment by LegalizeAdulthood** [comment]
@agauniyal Great, thanks.  Were the [getting started instructions](http://help.exercism.io/getting-started-with-cpp.html) missing something?

You mentioned that you had problems with CMake and boost libraries.  Without seeing the exact error or knowing the state of your system, it's hard for me to know exactly what went wrong.  You don't explicitly say what kind of operating system you are using, but I am guessing some linux distribution from what you've said.  Unfortunately the package manager situation for boost libraries is a bit of a mess in the linux world.  The prepackaged binaries can cause just as many problems as they solve when you use a compiler that is C++11 capable.  Unfortunately, this is out of my control.

Since I wrote the initial documentation, exercism.io has implemented submission of multiple files, but I haven't had a chance to test it yet in the C++ language track.  However, having said that it seems you got yourself into a bind because you duplicated code between the header file and the `.cpp` file resulting in multiply defined symbols.  What I did as a workaround in the meantime was to include the contents of my [header file as a comment](http://exercism.io/submissions/7c670eac9163260ded704dae).  This was suggested in the beginning of the help instructions:

> When using a .cpp file for the implementation, it is helpful to reviewers if the contents of the .h file ppear in the .cpp file at the top as a comment or #if disabled block.

The tests are going to include your headers, so the headers must contain the declarations.  If you want to put everything into a single file, then put the implementation in the header file as well.  This was also mentioned in the help, but rereading it now I can see how it could be parsed differently and I will tweak that part of the documentation to be more explicit about splitting things between the header `.h` and the source `.cpp`.

Regarding the use of Boost, I'm afraid that's just not something that is going to change.  It's too confusing to switch test frameworks in the middle and there are many things from Boost that are going to help you improve your code, even for [Bob](http://exercism.io/submissions/53c6c41aad807561e896113c).

Where you could help me improve this whole situation is to let me know which parts of the documentation weren't clear enough or were confusing so that they can be improved.

**Comment by agauniyal** [comment]
Coming from Ubuntu , I was somewhat used to it and was confused why cmake complaining. You can include these in getting started with cpp guide , for arch users : 

1. https://www.archlinux.org/packages/extra/x86_64/boost/
2. https://www.archlinux.org/packages/extra/x86_64/boost-libs/

so we can do `pacman -S boost boost-libs`. I was missing either of them and thought it would work.

And I guess writing seperate `.h` and `.cpp` to pass tests and then concatenating .h into .cpp as comments on top of the file for submitiing would be good way to do that.

Regarding boost , my only argument was the amount of errors it generated was a little off-putting at first. It generated over 10000s lines of errors and niether of them said a single thing about my mistake. I had to run static analysis over my code to find that error. Anyways thanks for quick support ,

**Comment by agauniyal** [comment]
also I didn't knew about boost trim function so that's cool. However we can do it within STL too , something like this?

    #include <iostream>
    #include <string>
    #include <algorithm>

    int main() {
	
	    std::string s = "  HE  L LO   WORL  D  1  ";
	    while(s.find(' ') != std::string::npos){
		    s.erase(std::find(s.begin() , s.end() , ' '));
	    }
	    std::cout<<s;
	    return 0;
    }

I haven't tested it yet , but I guess it should probably work.

**Comment by LegalizeAdulthood** [comment]
@agauniyal I looked at [your gist](https://gist.github.com/agauniyal/37b23320dd624b20941b) again, and while I can't really do anything about the amount of spew issued by a compiler when you make some mistake, I can offer general advice on how to deal with C++ compiler errors.

1. Don't panic!  The size of the error can be intimidating at first.
1. Always start with the first error.  Subsequent errors are often side-effects caused by the first error because the compiler was thrown off track and now everything afterwards is wrong.
1. If you don't understand the error, look for the first location mentioned in one of *your* source files for the source of the problem.  The error is not going to be in system headers, boost headers, or in the supplied unit test file.
1. If you still can't see the source of the problem, focus your code down to the minimum needed by commenting out chunks of code until the disappears.

Following this advice, we start by ignoring everything after the first error on line 24.  We also ignore everything talking about files we didn't write, including `bob_test.cpp`.  This means we need to focus on files we wrote, in this case the file `bob.cpp`.  (You must have modified the file `bob_test.cpp`, because `bob_test.cpp` does not directly include `bob.cpp`, but both files are mentioned on lines 2-3.)  Having done this, we see that the compiler is complaining about things happening around a `{` token.  I don't know what your code looked like when you got this error, so I can't say specifically how to apply the information from the error to your code.

However, from your description of your process, you weren't incrementally developing your code in the exercism project but were instead developing it in the online IDE ideone.  While ideone is great for experimenting, I would not recommend trying to develop your solution there and bringing it back to the tests because the simple act of building your code up element by element in conjunction with the tests helps you to avoid exactly these kinds of mistakes.

In C++, test frameworks are necessarily implemented by means of macros and macros can lead to such problems as this giant error message.  This is why the getting started documentation links to my C++ Now! 2014 course on TDD with C++ and Boost.Test because all these details are explained there.  If you haven't yet, I recommend you look at the presentation with that course because it really does go step-by-step through the process of creating an implementation in response to writing tests.  When you build up code slowly in small steps, it avoids the psychological shock of taking what you think is a working solution from ideone and then pasting it into the exercism project and expecting all the tests to pass but instead being shotgun blasted to the face with a giant spew of errors.  Always work by taking [small moves](https://www.youtube.com/watch?v=oB6NNbFHjCc).

I will add the above advice on error messages to the getting started documentation.  This understanding of how to work with the compiler and understand error messages was assumed as basic knowledge for a C++ programmer.  It's one of those things where it's hard to "unlearn" what you already know when explaining things to someone who doesn't already know them.
